pkg/validation/validate_component.go

```diff
@@ -47,75 +50,106 @@ func NewComponentValidator(c []*bundle.ComponentPackage, set *bundle.ComponentSe
 }
 
 // Validate validates components and components sets, providing as many errors as it can.
-func (v *ComponentValidator) Validate() []error {
-	var errs []error
-	errs = append(errs, v.validateComponentsSet()...)
+func (v *ComponentValidator) Validate() field.ErrorList {
+	errs := field.ErrorList{}
+	errs = append(errs, v.validateComponentSet()...)
 	errs = append(errs, v.validateComponentPackages()...)
 	errs = append(errs, v.validateObjects()...)
 	return errs
 }
 
-func (v *ComponentValidator) validateComponentPackages() []error {
-	var errs []error
+func cPath(ref bundle.ComponentReference) *field.Path {
+	return field.NewPath(fmt.Sprintf("Component{%s, %s}", ref.ComponentName, ref.Version))
+}
+
+func (v *ComponentValidator) validateComponentPackages() field.ErrorList {
+	errs := field.ErrorList{}
 	objCollect := make(map[bundle.ComponentReference]bool)
-	for _, ca := range v.components {
+	for i, ca := range v.components {
+		pi := field.NewPath("Component").Index(i)
+
 		n := ca.Spec.ComponentName
-		if err := validateName(n); err != nil {
-			errs = append(errs, fmt.Errorf("the component name %q was invalid config %v", n, ca))
+		if n == "" {
+			errs = append(errs, field.Required(pi.Child("Spec", "ComponentName"), ""))
+		}
+		ver := ca.Spec.Version
+		if ver == "" {
+			errs = append(errs, field.Required(pi.Child("Spec", "Version"), ""))
+		}
+		if n == "" || ver == "" {
+			// Subsequent validation relies on components having a unique name+version pair.
+			continue
+		}
+
+		ref := ca.ComponentReference()
+		p := cPath(ref)
+
+		if nameErrs := validateName(p.Child("Spec", "ComponentName"), n); len(errs) > 0 {
+			errs = append(errs, nameErrs...)
 		}
 
 		api := ca.APIVersion
 		if !apiVersionPattern.MatchString(api) {
-			errs = append(errs, fmt.Errorf("components APIVersion must have an apiVersion of the form \"bundle.gke.io/<version>\". was %q for config named %q", api, n))
+			errs = append(errs, field.Invalid(p.Child("APIVersion").Index(i), api, "must have the form \"bundle.gke.io/<version>\""))
```

Comment:
Could use backticks not have to escape the quotes:
```
`must have the form "bundle.gke.io/<version>"`
```

Reply:
Done